File: 02/src/lexer.c
#include "emolang.h"


// 包含所有 34 個 Emoji 的完整關鍵字表
const char *emoji_keywords[] = {
    "📦", "🔢", "🎈", "📝", "🚦", "🟰", "🤔", "🤷", "🔁", "👇", 
    "👆", "📢", "➕", "➖", "✖️", "➗", "✂️", "🤝", "📈", "📉", 
    "🎡", "🚧", "🏗️", "🆕", "➡️", "📍", "🎯", "📚", "📌", "📥", 
    "🟢", "🔴", "🛠️", "🔙", "🔗", "🔀", "🙅",
    "📋", "📖", "🛒", "📏"
};
// 對應的 34 個 Token
TokenType emoji_tokens[] = {
    TOK_LET, TOK_LET, TOK_LET, TOK_LET, TOK_LET,
    TOK_ASSIGN, TOK_IF, TOK_ELSE, TOK_WHILE, TOK_LBRACE, TOK_RBRACE, TOK_PRINT, TOK_PLUS, TOK_MINUS, TOK_MUL, 
    TOK_DIV, TOK_MOD, TOK_EQ, TOK_GT, TOK_LT, TOK_FOR, TOK_SEP, TOK_STRUCT, TOK_NEW, TOK_DOT, 
    TOK_REF, TOK_DEREF, TOK_ARRAY, TOK_INDEX, TOK_INPUT, TOK_TRUE, TOK_FALSE, TOK_FUNC, TOK_RETURN,
    TOK_AND, TOK_OR, TOK_NOT,
    TOK_LIST, TOK_DICT, TOK_APPEND, TOK_LEN
};

int match_keyword(const char *p) {
    for (int i = 0; i < 41; i++) { 
        int len = strlen(emoji_keywords[i]);
        if (strncmp(p, emoji_keywords[i], len) == 0) return i;
    }
    return -1;
}

void advance_token() {
    // 略過空白字元
    while (isspace(src_code[src_pos])) src_pos++;
    
    // 檢查是否結束
    if (src_code[src_pos] == '\0') { current_token.type = TOK_EOF; return; }
    
    // 標點符號判斷 (支援括號與函數參數的逗號)
    if (src_code[src_pos] == '(') { current_token.type = TOK_LPAREN; strcpy(current_token.value, "("); src_pos++; return; }
    if (src_code[src_pos] == ')') { current_token.type = TOK_RPAREN; strcpy(current_token.value, ")"); src_pos++; return; }
    if (src_code[src_pos] == ',') { current_token.type = TOK_COMMA; strcpy(current_token.value, ","); src_pos++; return; }
    
    // 解析字串
    if (src_code[src_pos] == '"') {
        src_pos++; int i = 0;
        while (src_code[src_pos] != '"' && src_code[src_pos] != '\0' && i < 255) {
            current_token.value[i++] = src_code[src_pos++];
        }
        current_token.value[i] = '\0';
        if (src_code[src_pos] == '"') src_pos++;
        current_token.type = TOK_STR; return;
    }
    
    // 解析數字與小數
    if (isdigit(src_code[src_pos])) {
        int i = 0, is_float = 0;
        while (isdigit(src_code[src_pos]) || src_code[src_pos] == '.') {
            if (src_code[src_pos] == '.') is_float = 1;
            current_token.value[i++] = src_code[src_pos++];
        }
        current_token.value[i] = '\0';
        current_token.type = is_float ? TOK_FLOAT_NUM : TOK_NUM; return;
    }
    
    // 解析 Emoji 關鍵字
    int kw_idx = match_keyword(&src_code[src_pos]);
    if (kw_idx != -1) {
        current_token.type = emoji_tokens[kw_idx];
        strcpy(current_token.value, emoji_keywords[kw_idx]);
        src_pos += strlen(emoji_keywords[kw_idx]); return;
    }
    
    // 解析普通變數名稱
    int i = 0;
    while (src_code[src_pos] != '\0' && !isspace(src_code[src_pos]) && 
           match_keyword(&src_code[src_pos]) == -1 && src_code[src_pos] != '"' &&
           src_code[src_pos] != '(' && src_code[src_pos] != ')' && src_code[src_pos] != ',') {
        current_token.value[i++] = src_code[src_pos++];
    }
    current_token.value[i] = '\0';
    current_token.type = TOK_ID;
}

void eat(TokenType type) {
    if (current_token.type == type) advance_token();
    else { my_error("語法錯誤: 遇到未預期的 Token", current_token.value); }
}

File: 03/emolang/src/lexer.py
from tokens import TokenType, Token


class EmoLangLexer:
    def __init__(self, src):
        self.src = src
        self.pos = 0
        self.current_token = None

        self.emoji_keywords = {
            "📦": TokenType.TOK_LET,
            "🔢": TokenType.TOK_LET,
            "🎈": TokenType.TOK_LET,
            "📝": TokenType.TOK_LET,
            "🚦": TokenType.TOK_LET,
            "🟰": TokenType.TOK_ASSIGN,
            "🤔": TokenType.TOK_IF,
            "🤷": TokenType.TOK_ELSE,
            "🔁": TokenType.TOK_WHILE,
            "👇": TokenType.TOK_LBRACE,
            "👆": TokenType.TOK_RBRACE,
            "📢": TokenType.TOK_PRINT,
            "➕": TokenType.TOK_PLUS,
            "➖": TokenType.TOK_MINUS,
            "✖️": TokenType.TOK_MUL,
            "➗": TokenType.TOK_DIV,
            "✂️": TokenType.TOK_MOD,
            "🤝": TokenType.TOK_EQ,
            "📈": TokenType.TOK_GT,
            "📉": TokenType.TOK_LT,
            "🎡": TokenType.TOK_FOR,
            "🚧": TokenType.TOK_SEP,
            "🏗️": TokenType.TOK_STRUCT,
            "🆕": TokenType.TOK_NEW,
            "➡️": TokenType.TOK_DOT,
            "📍": TokenType.TOK_REF,
            "🎯": TokenType.TOK_DEREF,
            "📚": TokenType.TOK_ARRAY,
            "📌": TokenType.TOK_INDEX,
            "📥": TokenType.TOK_INPUT,
            "🟢": TokenType.TOK_TRUE,
            "🔴": TokenType.TOK_FALSE,
            "🛠️": TokenType.TOK_FUNC,
            "🔙": TokenType.TOK_RETURN,
            "🔗": TokenType.TOK_AND,
            "🔀": TokenType.TOK_OR,
            "🙅": TokenType.TOK_NOT,
        }

    def skip_space(self):
        while self.pos < len(self.src) and self.src[self.pos].isspace():
            self.pos += 1

    def match_keyword(self):
        for kw in self.emoji_keywords:
            if self.src.startswith(kw, self.pos):
                return kw
        return None

    def advance(self):
        self.skip_space()
        if self.pos >= len(self.src):
            self.current_token = Token(TokenType.TOK_EOF, "")
            return

        if self.src[self.pos] == '(':
            self.current_token = Token(TokenType.TOK_LPAREN, "(")
            self.pos += 1
            return
        if self.src[self.pos] == ')':
            self.current_token = Token(TokenType.TOK_RPAREN, ")")
            self.pos += 1
            return
        if self.src[self.pos] == ',':
            self.current_token = Token(TokenType.TOK_COMMA, ",")
            self.pos += 1
            return

        if self.src[self.pos] == '"':
            self.pos += 1
            value = ""
            while self.pos < len(self.src) and self.src[self.pos] != '"':
                value += self.src[self.pos]
                self.pos += 1
            if self.pos < len(self.src):
                self.pos += 1
            self.current_token = Token(TokenType.TOK_STR, value)
            return

        if self.src[self.pos].isdigit():
            is_float = False
            value = ""
            while self.pos < len(self.src) and (self.src[self.pos].isdigit() or self.src[self.pos] == '.'):
                if self.src[self.pos] == '.':
                    is_float = True
                value += self.src[self.pos]
                self.pos += 1
            self.current_token = Token(TokenType.TOK_FLOAT_NUM if is_float else TokenType.TOK_NUM, value)
            return

        kw = self.match_keyword()
        if kw:
            self.current_token = Token(self.emoji_keywords[kw], kw)
            self.pos += len(kw)
            return

        value = ""
        while self.pos < len(self.src) and not self.src[self.pos].isspace():
            if self.match_keyword():
                break
            if self.src[self.pos] in '(),':
                break
            value += self.src[self.pos]
            self.pos += 1
        self.current_token = Token(TokenType.TOK_ID, value)

    def eat(self, expected_type):
        if self.current_token.type == expected_type:
            self.advance()
        else:
            raise RuntimeError(f"語法錯誤: 期待 {expected_type}，但遇到 {self.current_token.type}")